### DISCUSSION TITLE
using uvicorn inside my script with reload=True results in "Address already in use"

### DISCUSSION THREAD

**Discussion by tropxy** [Kludex/uvicorn#1854] [Potential Issue] [answered]
Hi, folks!
This is my app template:

```python
import logging
import uvicorn
from fastapi import FastAPI
from pydantic import BaseModel


# FIXME: Set up logging configuration in a separate file
logging.basicConfig(level=logging.DEBUG)
# Avoid Spamming
logging.getLogger("asyncio").setLevel(logging.WARNING)

logger = logging.getLogger("app_logger")


class RegistrationData(BaseModel):
    username: str
    password: str

class RegisteredUser(BaseModel):
    id: int

app = FastAPI()

@app.post("/", response_model=RegisteredUser)
async def register(user: RegistrationData):
    from asyncio import sleep
    logger.debug(f"Registration data {user}")
    await sleep(0.5)
    return RegisteredUser(id=5)

@app.get("/")
async def hello():
    return "hello there"


def run():
    """
    Serving the app
    """
    uvicorn.run(
        __name__ + ':app',
        reload=True,
        host='0.0.0.0',
        port=8080,
        log_level='debug',
    )

```

This results in the following debug message:
> INFO:     Will watch for changes in these directories: ['/Users/andre/Devel/switch_gridopio']
INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
INFO:     Started reloader process [74906] using StatReload
INFO:     Will watch for changes in these directories: ['/Users/andre/Devel/switch_gridopio']
ERROR:    [Errno 48] Address already in use
INFO:     Stopping reloader process [74906]

And I cant load the docs or the redocs, however If I set the reload to "False" then it works. Any thoughts? 
Thank you!

**Comment by duyker** [comment]
I am not totally sure _why_ but it works for when I remove the `host='0.0.0.0'` value:

```python
uvicorn.run(
    "main:app",
    host="0.0.0.0", # REMOVE THIS LINE
    port=5000,
    reload=True,
    server_header=False,
)
```

**Comment by Kludex** [comment]
What OS are you using?

**Reply by duyker** [reply]
macOS Ventura 13.3 (Apple M1)

> NOTE: the values `localhost` and `127.0.0.1` work fine, its just `0.0.0.0` that does not work.

**Reply by duyker** [reply, marked as answer]
There is a bit of guess work involved here, but I found a link that pointed to the real problem.

It looks like something else was running on port `:5000` (_AirPlay Receiver_ to be exact).

When specifying `localhost` or `127.0.0.1` it seems to hijack the port (the FastAPI receives requests fine). It seems to only indicate that there is a port conflict if you specify `0.0.0.0`.

Picking another port that was not in conflict allowed me to use the `host='0.0.0.0'` without issues.

**Reference:**

https://progressstory.com/tech/port-5000-already-in-use-macos-monterey-issue/

**Comment by MichaelTsao** [comment]
If you're running programmatically, using uvicorn.run(...), then use equivalent keyword arguments, eg. uvicorn.run("example:app", port=5000, reload=True, access_log=False). Please note that in this case, if you use reload=True or workers=NUM, you should put uvicorn.run into if __name__ == '__main__' clause in the main module.

https://www.uvicorn.org/settings/
